File: android/app/src/main/kotlin/com/example/expense_app/TextRecognitionHandler.kt
package com.example.expense_app

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Matrix
import android.util.Base64
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.TextRecognizer
import com.google.mlkit.vision.text.latin.TextRecognizerOptions

class TextRecognitionHandler(private val recognizer: TextRecognizer) {

    fun processImage(byteArray: ByteArray, rotation: Int = 0, result: (Exception?, Map<String, Any>?) -> Unit) {
        val image = InputImage.fromBitmap(getBitmap(byteArray, rotation), 0)
        recognizer.process(image).addOnSuccessListener {
            result.invoke(null, TextMapper.toMap(it))
        }.addOnFailureListener {
            result.invoke(it, null)
        }
    }

    private fun getBitmap(byteArray: ByteArray, rotation: Int): Bitmap {
        return rotateBitmap(BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size), rotation)
    }

    private fun rotateBitmap(bitmap: Bitmap, rotation: Int): Bitmap {
        val matrix = Matrix()
        matrix.postRotate(rotation.toFloat())
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }

}

File: ios/Runner/TextRecognitionHandler.swift
//
//  TextRecognitionHandler.swift
//  Runner
//
//  Created by Fikri Haikal on 31/01/23.
//

import Foundation
import MLKit
import CoreGraphics
import Alamofire

class TextRecognitionHandler {
    
    let textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
    
    func processImage(byteArray:  [UInt8], rotation: CGFloat, resulText: @escaping (String?, [String: Any]?) -> Void) {
        
        // Convert byte array to UIImage
        guard let originalImage = UIImage(data: Data(byteArray)) else {
            resulText("error", nil)
            return
        }

        // Rotate the image if necessary
        let rotatedImage = rotateImage(image: originalImage, degrees: rotation)

        // Create a VisionImage from the rotated image
        let visionImage = VisionImage(image: rotatedImage)
        visionImage.orientation = .up

        // Process the image with the text recognizer
        textRecognizer.process(visionImage) { result, error in
            guard error == nil, let result = result else {
                resulText("error", nil)
                return
            }
            var mappedResult = TextMapper.toMap(text: result)
            // Convert the result to a Flutter map and return it
            resulText(nil, mappedResult)
        }
        
    }
    
    func convertImageToBase64(image: UIImage) -> String? {
        if let imageData = image.jpegData(compressionQuality: 0.1) {
            let base64String = imageData.base64EncodedString(options: Data.Base64EncodingOptions.lineLength64Characters)
            return base64String
        }
        return nil
    }
    
    func rotateImage(image: UIImage, degrees: CGFloat) -> UIImage {
        // Calculate the radians needed to rotate the image
        let radians = degrees * CGFloat.pi / 180.0
        
        // Set the size of the rotated image
        let rotatedSize = CGRect(origin: .zero, size: image.size)
            .applying(CGAffineTransform(rotationAngle: radians))
            .integral.size
        
        // Create a context with the rotated size
        UIGraphicsBeginImageContext(rotatedSize)
        
        // Rotate the context around the center point
        let context = UIGraphicsGetCurrentContext()!
        context.translateBy(x: rotatedSize.width / 2.0, y: rotatedSize.height / 2.0)
        context.rotate(by: radians)
        
        // Draw the image in the context
        image.draw(in: CGRect(x: -image.size.width / 2.0, y: -image.size.height / 2.0,
                              width: image.size.width, height: image.size.height))
        
        // Create a new image from the context
        let rotatedImage = UIGraphicsGetImageFromCurrentImageContext()!
        UIGraphicsEndImageContext()
        
        return rotatedImage
    }

    func sendBase64ToAPI(image: UIImage) {
        let base64 = convertImageToBase64(image: image)
        
        let url = "http://103.250.11.69:9000/api/send-dummy"
        let parameters: [String: Any] = [
            "message": base64,
        ]
        let headers: HTTPHeaders = [
            "Content-Type": "application/json"
        ]
        
        AF.request(url, method: .post, parameters: parameters, encoding: JSONEncoding.default, headers: headers).responseJSON { response in
            switch response.result {
            case .success(let value):
                print("API request successful with response: \(value)")
                // Handle the response data here
            case .failure(let error):
                print("API request failed with error: \(error)")
                // Handle the error here
            }
        }
    }

}
